mktables: Fix Useless use of const in void context

khwilliamson opened the pull request:
The code was attempting to get a slice of an entire array excluding the
0th element.  But it was using the wrong incantation.

The array is the path to a file with each element being one directory.

And it just happened to work because of the current layout of the
directories.

This commit moves this slice fetch so it is not done in the loop, adds
another sanity check, and it is generally clearer what is going on.

This closes #23520 

* This set of changes does not require a perldelta entry.

karenetheridge commented on lib/unicore/mktables:
```diff
@@ -16087,7 +16087,10 @@ sub register_file_for_name($table, $directory_ref, $file) {
-        my $sub_filename = join('/', $directory_ref->[1, -1], $file);
+        # Get the path to $file stripping off any top level directory
+        my $sub_filename = join('/',
+                                ($directory_ref->@[1, $#$directory_ref] // ""),
```
Is $directory_ref always three elements?

If not, then I think we would want a range operator here (`1 .. $#$directory_ref`) instead?

And if it's always three elements, we could shorten this to just `$directory_ref->@[1..2]`.

khwilliamson replied:
I was working too hastily; now reworked

tonycoz commented on lib/unicore/mktables:
```diff
@@ -16087,7 +16087,10 @@ sub register_file_for_name($table, $directory_ref, $file) {
-        my $sub_filename = join('/', $directory_ref->[1, -1], $file);
+        # Get the path to $file stripping off any top level directory
+        my $sub_filename = join('/',
+                                ($directory_ref->@[1, $#$directory_ref] // ""),
```
There's a couple of problems with:
```
($directory_ref->@[1, $#$directory_ref] // "")
```
- the `//` forces the result into scalar context, so only the second element is returned
- if we remove the `//` then only the second and last elements of the array are returned, and if the array has two elements they may be the same element.

Should this be:
```
$directory_ref->@[1 ... $#$directory_ref]
```
?

```
tony@venus:.../git/perl6$ perl ../23542.pl
path 'a/b/c/d/e'
orig 'e/file1'
new  'e/file1'
wrong'b/e/file1'
fix? 'b/c/d/e/file1'
path 'a/b/c'
orig 'c/file1'
new  'c/file1'
wrong'b/c/file1'
fix? 'b/c/file1'
path 'a/b'
orig 'b/file1'
new  'b/file1'
wrong'b/b/file1'
fix? 'b/file1'
tony@venus:.../git/perl6$ cat ../23542.pl
use v5.36;

my $file = "file1";
for my $path ("a/b/c/d/e", "a/b/c", "a/b") {
  my $directory_ref = [ split /\//, $path ];
  say "path '$path'";
  say "orig '", join('/', $directory_ref->[1, -1], $file), "'";
  say "new  '", join('/', ($directory_ref->@[1, $#$directory_ref] // ""), $file), "'";
  say "wrong'", join('/', $directory_ref->@[1, $#$directory_ref], $file), "'";
  say "fix? '", join('/', $directory_ref->@[1 .. $#$directory_ref], $file), "'";
}
```

khwilliamson replied:
Sorry; reworked

khwilliamson committed:
mktables: Correct comment

Spotted by @bram-perl

khwilliamson committed:
mktables: Fix Useless use of const in void context

The code was attempting to get a slice of an entire array excluding the
0th element.  But it was using the wrong incantation.

The array is the path to a file with each element being one directory.

And it just happened to work because of the current layout of the
directories.

This commit moves this slice fetch so it is not done in the loop, adds
another sanity check, and is generally clearer what is going on.

tonycoz commented on lib/unicore/mktables:
```diff
@@ -16072,9 +16072,18 @@ sub register_file_for_name($table, $directory_ref, $file) {
     # pseudo-directory '#'
-    if ($directory_ref->[0] !~ / ^ $matches_directory | \# $ /x) {
+    my $sub_directory;
+    if ($directory_ref->[0] =~ /^#$/) {
```
can `$directory_ref->[0]` ever be `"#\n"` ?

If not, why use a regexp rather than `eq`?

khwilliamson replied:
I wasn't thinking
